vignettes/decorate-module-output.Rmd

```diff
@@ -0,0 +1,513 @@
+---
+title: "Customizing Module Output"
+author: "NEST CoreDev"
+output:
+  rmarkdown::html_vignette:
+    toc: true
+vignette: >
+  %\VignetteIndexEntry{Customizing Module Output}
+  %\VignetteEngine{knitr::rmarkdown}
+  %\VignetteEncoding{UTF-8}
+---
+
+## Introduction
+
+`teal` is a powerful `shiny`-based framework with built-in modules for interactive data analysis. 
+This document outlines the customization options available for modifying the output of `teal` modules. 
+You will learn how to use [`teal_transform_module`](https://insightsengineering.github.io/teal/latest-tag/reference/teal_transform_module.html) to modify and enhance the objects created by `teal::modules()`, 
+enabling you to tailor the outputs to your specific requirements without rewriting the original module code.
+
+## Decorators
+
+In programming, **decoration** refers to the process of modifying an object while preserving its original class. For instance, given an object `x` of class `"my_class"`, a function `foo(x)` is considered a **decorator function** if it modifies `x` and returns an object that retains the same class. In this context, `x` is referred to as the **decorated object**, and `foo()` is the **decorator function** or **decorator**. Decorators can perform a variety of operations, such as adding new methods or modifying data, while ensuring the object remains compatible with its original usage.
+
+In the context of `teal` applications, decoration is specifically used to modify module outputs, such as plots or tables. For example, consider a decorator function `add_title(x, <plot_title>)` that takes a `ggplot2` plot object (`x`) as input, applies a title modification, and returns a modified `ggplot2` plot object. This function qualifies as a decorator because it preserves the original class of the input object. Conversely, a function like `create_plot(<data>, <x_axis>, <y_axis>)`, which generates a new plot object, is **not** a decorator, as it produces an output of a different class.
+
+Preserving the object's class during decoration is essential for compatibility. It ensures that the subsequent "display" logic can seamlessly handle both decorated and non-decorated objects.
+
+The decoration process can vary in complexity:
```

Comment:
This list doesn't show up nicely on the browser on my computer:

![screenshot of this paragraph](https://github.com/user-attachments/assets/6b643421-5cd6-4a22-a59c-f72b1ca6a9ea)

I think it needs a new line before it:

```suggestion
The decoration process can vary in complexity:

```

Reply:
Good catch!

My assumption was that this is automatically detected - it seems that it's not

<details>
  <summary>markdownlint results</summary>

  ```
❯ markdownlint vignettes/decorate-module-output.Rmd
vignettes/decorate-module-output.Rmd:15:98 MD009/no-trailing-spaces Trailing spaces [Expected: 0 or 2; Actual: 1]
vignettes/decorate-module-output.Rmd:15:81 MD013/line-length Line length [Expected: 80; Actual: 98]
vignettes/decorate-module-output.Rmd:16:103 MD009/no-trailing-spaces Trailing spaces [Expected: 0 or 2; Actual: 1]
vignettes/decorate-module-output.Rmd:16:81 MD013/line-length Line length [Expected: 80; Actual: 103]
vignettes/decorate-module-output.Rmd:17:208 MD009/no-trailing-spaces Trailing spaces [Expected: 0 or 2; Actual: 1]
vignettes/decorate-module-output.Rmd:17:81 MD013/line-length Line length [Expected: 80; Actual: 208]
vignettes/decorate-module-output.Rmd:18:81 MD013/line-length Line length [Expected: 80; Actual: 108]
vignettes/decorate-module-output.Rmd:22:81 MD013/line-length Line length [Expected: 80; Actual: 587]
vignettes/decorate-module-output.Rmd:24:81 MD013/line-length Line length [Expected: 80; Actual: 597]
vignettes/decorate-module-output.Rmd:26:81 MD013/line-length Line length [Expected: 80; Actual: 190]
vignettes/decorate-module-output.Rmd:29:81 MD013/line-length Line length [Expected: 80; Actual: 120]
vignettes/decorate-module-output.Rmd:29 MD032/blanks-around-lists Lists should be surrounded by blank lines [Context: "- **Simple Decorations**: Sing..."]
vignettes/decorate-module-output.Rmd:30:81 MD013/line-length Line length [Expected: 80; Actual: 161]
vignettes/decorate-module-output.Rmd:32:81 MD013/line-length Line length [Expected: 80; Actual: 290]
vignettes/decorate-module-output.Rmd:38:81 MD013/line-length Line length [Expected: 80; Actual: 222]
vignettes/decorate-module-output.Rmd:39:81 MD013/line-length Line length [Expected: 80; Actual: 272]
vignettes/decorate-module-output.Rmd:41:81 MD013/line-length Line length [Expected: 80; Actual: 136]
vignettes/decorate-module-output.Rmd:45:109 MD009/no-trailing-spaces Trailing spaces [Expected: 0 or 2; Actual: 1]
vignettes/decorate-module-output.Rmd:45:81 MD013/line-length Line length [Expected: 80; Actual: 109]
vignettes/decorate-module-output.Rmd:48:81 MD013/line-length Line length [Expected: 80; Actual: 116]
vignettes/decorate-module-output.Rmd:49:110 MD009/no-trailing-spaces Trailing spaces [Expected: 0 or 2; Actual: 1]
vignettes/decorate-module-output.Rmd:49:81 MD013/line-length Line length [Expected: 80; Actual: 110]
vignettes/decorate-module-output.Rmd:50:213 MD009/no-trailing-spaces Trailing spaces [Expected: 0 or 2; Actual: 1]
vignettes/decorate-module-output.Rmd:50:81 MD013/line-length Line length [Expected: 80; Actual: 213]
vignettes/decorate-module-output.Rmd:51:81 MD013/line-length Line length [Expected: 80; Actual: 98]
vignettes/decorate-module-output.Rmd:55:81 MD013/line-length Line length [Expected: 80; Actual: 227]
vignettes/decorate-module-output.Rmd:56:74 MD009/no-trailing-spaces Trailing spaces [Expected: 0 or 2; Actual: 1]
vignettes/decorate-module-output.Rmd:57:106 MD009/no-trailing-spaces Trailing spaces [Expected: 0 or 2; Actual: 1]
vignettes/decorate-module-output.Rmd:57:81 MD013/line-length Line length [Expected: 80; Actual: 106]
vignettes/decorate-module-output.Rmd:78:62 MD009/no-trailing-spaces Trailing spaces [Expected: 0 or 2; Actual: 1]
vignettes/decorate-module-output.Rmd:79:73 MD009/no-trailing-spaces Trailing spaces [Expected: 0 or 2; Actual: 1]
vignettes/decorate-module-output.Rmd:80:81 MD013/line-length Line length [Expected: 80; Actual: 202]
vignettes/decorate-module-output.Rmd:96:144 MD009/no-trailing-spaces Trailing spaces [Expected: 0 or 2; Actual: 1]
vignettes/decorate-module-output.Rmd:96:81 MD013/line-length Line length [Expected: 80; Actual: 144]
vignettes/decorate-module-output.Rmd:97:126 MD009/no-trailing-spaces Trailing spaces [Expected: 0 or 2; Actual: 1]
vignettes/decorate-module-output.Rmd:97:81 MD013/line-length Line length [Expected: 80; Actual: 126]
vignettes/decorate-module-output.Rmd:98:81 MD013/line-length Line length [Expected: 80; Actual: 180]
vignettes/decorate-module-output.Rmd:126:209 MD009/no-trailing-spaces Trailing spaces [Expected: 0 or 2; Actual: 1]
vignettes/decorate-module-output.Rmd:126:81 MD013/line-length Line length [Expected: 80; Actual: 209]
vignettes/decorate-module-output.Rmd:127:81 MD013/line-length Line length [Expected: 80; Actual: 193]
vignettes/decorate-module-output.Rmd:149:80 MD009/no-trailing-spaces Trailing spaces [Expected: 0 or 2; Actual: 1]
vignettes/decorate-module-output.Rmd:150:81 MD013/line-length Line length [Expected: 80; Actual: 123]
vignettes/decorate-module-output.Rmd:151:81 MD013/line-length Line length [Expected: 80; Actual: 133]
vignettes/decorate-module-output.Rmd:152:81 MD013/line-length Line length [Expected: 80; Actual: 119]
vignettes/decorate-module-output.Rmd:182:81 MD013/line-length Line length [Expected: 80; Actual: 123]
vignettes/decorate-module-output.Rmd:183:81 MD013/line-length Line length [Expected: 80; Actual: 134]
vignettes/decorate-module-output.Rmd:205:120 MD009/no-trailing-spaces Trailing spaces [Expected: 0 or 2; Actual: 1]
vignettes/decorate-module-output.Rmd:205:81 MD013/line-length Line length [Expected: 80; Actual: 120]
vignettes/decorate-module-output.Rmd:206:80 MD009/no-trailing-spaces Trailing spaces [Expected: 0 or 2; Actual: 1]
vignettes/decorate-module-output.Rmd:210 MD012/no-multiple-blanks Multiple consecutive blank lines [Expected: 1; Actual: 2]
vignettes/decorate-module-output.Rmd:212:81 MD013/line-length Line length [Expected: 80; Actual: 93]
vignettes/decorate-module-output.Rmd:312 MD024/no-duplicate-heading Multiple headings with the same content [Context: "Example Module"]
vignettes/decorate-module-output.Rmd:314:81 MD013/line-length Line length [Expected: 80; Actual: 308]
vignettes/decorate-module-output.Rmd:406:81 MD013/line-length Line length [Expected: 80; Actual: 87]
vignettes/decorate-module-output.Rmd:421 MD012/no-multiple-blanks Multiple consecutive blank lines [Expected: 1; Actual: 2]
vignettes/decorate-module-output.Rmd:493 MD024/no-duplicate-heading Multiple headings with the same content [Context: "Application"]
  ```
</details>

Your observation is there - among tons of other things :)
[markdownlint](https://github.com/igorshubovych/markdownlint-cli) is designed for `.md` files. I believe that `.Rmd` should have a similar structure (if not identical) - worst case we would have to disable some of the rules. I'm thinking that we should configure our superlinter CI workflow to process `.Rmd` with `markdownlint`. Let me create an issue for this. For the time being - I would recommend to test it manually.

Reply:
https://github.com/insightsengineering/idr-tasks/issues/712

Reply:
Nice, the heading checks might help to structure better the vignettes or simplify the sections. 
I think you meant to link to https://github.com/insightsengineering/idr-tasks/issues/823